src/components/cuda/cupti_profiler.c

```diff
@@ -0,0 +1,1564 @@
+    NVPW_CounterDataBuilder_CreatePtr                 = NULL;
+    NVPW_CounterDataBuilder_DestroyPtr                = NULL;
+    NVPW_CounterDataBuilder_AddMetricsPtr             = NULL;
+    NVPW_CounterDataBuilder_GetCounterDataPrefixPtr   = NULL;
+    NVPW_CounterData_GetNumRangesPtr                  = NULL;
+    NVPW_Profiler_CounterData_GetRangeDescriptionsPtr = NULL;
+    NVPW_MetricsContext_SetCounterDataPtr             = NULL;
+    NVPW_MetricsContext_EvaluateToGpuValuesPtr        = NULL;
+    NVPW_RawMetricsConfig_GetNumPassesPtr             = NULL;
+    NVPW_RawMetricsConfig_SetCounterAvailabilityPtr   = NULL;
+    NVPW_RawMetricsConfig_IsAddMetricsPossiblePtr     = NULL;
+    NVPW_MetricsContext_GetCounterNames_BeginPtr      = NULL;
+    NVPW_MetricsContext_GetCounterNames_EndPtr        = NULL;
+    return PAPI_OK;
+}
+
+static int initialize_cupti_profiler_api(void)
+{
+    COMPDBG("Entering.\n");
+    int papi_errno;
+    CUpti_Profiler_Initialize_Params profilerInitializeParams = { CUpti_Profiler_Initialize_Params_STRUCT_SIZE, NULL };
+    papi_errno = cuptiProfilerInitializePtr(&profilerInitializeParams);
+    if (papi_errno != CUPTI_SUCCESS) {
+        ERRDBG("CUPTI error %d: cuptiProfilerInitialize failed.\n", papi_errno);
+        return PAPI_EMISC;
+    }
+    return PAPI_OK;
+}
+
+static int deinitialize_cupti_profiler_api(void)
+{
+    COMPDBG("Entering.\n");
+    int papi_errno;
+    CUpti_Profiler_DeInitialize_Params profilerDeInitializeParams = { CUpti_Profiler_DeInitialize_Params_STRUCT_SIZE, NULL };
+    papi_errno = cuptiProfilerDeInitializePtr(&profilerDeInitializeParams);
+    if (papi_errno != CUPTI_SUCCESS) {
+        ERRDBG("CUPTI Error %d: cuptiProfilerDeInitialize failed.\n", papi_errno);
+        return PAPI_EMISC;
+    }
+    return PAPI_OK;
+}
+
+static int initialize_perfworks_api(void)
+{
+    COMPDBG("Entering.\n");
+    int papi_errno;
+    NVPW_InitializeHost_Params perfInitHostParams = { NVPW_InitializeHost_Params_STRUCT_SIZE, NULL };
+    papi_errno = NVPW_InitializeHostPtr(&perfInitHostParams);
+    if (papi_errno != NVPA_STATUS_SUCCESS) {
+        ERRDBG("NVPW Error %d: NVPW_InitializeHostPtr failed.\n", papi_errno);
+        return PAPI_EMISC;
+    }
+    return PAPI_OK;
+}
+
+static int get_chip_name(int dev_num, char* chipName)
+{
+    int papi_errno;
+    CUpti_Device_GetChipName_Params getChipName = {
+        .structSize = CUpti_Device_GetChipName_Params_STRUCT_SIZE,
+        .pPriv = NULL,
+        .deviceIndex = 0
+    };
+    getChipName.deviceIndex = dev_num;
+    papi_errno = cuptiDeviceGetChipNamePtr(&getChipName);
+    if (papi_errno != CUPTI_SUCCESS) {
+        ERRDBG("CUPTI error %d: Failed to get chip name for device %d\n", papi_errno, dev_num);
+        return PAPI_EMISC;
+    }
+    strcpy(chipName, getChipName.pChipName);
+    return PAPI_OK;
+}
+
+struct byte_array_s {
+    int size;
+    uint8_t *data;
+};
+
+struct cuptip_gpu_control_s {
```

Comment:
`cuptip_gpu_control_s` and `cuptip_control_s` sounds kind of the same thing from the name. Maybe you could rename `cuptip_gpu_control_s` to something more distinctive? I don't really have a good name for that. Maybe you can think of something?

Reply:
I think that might work better.